### PR TITLE
feat: [FC-0047] Improved Dashboard Level Navigation

### PR DESCRIPTION
**Pull request by IvanStepanok**
## Quick Preview

https://github.com/openedx/openedx-app-ios/assets/128456094/4993cb6d-a9bb-4275-84bc-e3c75f1913a0

## Drop down menu
They are enabled automatically when "Programs" is enabled in the configuration file.

https://github.com/openedx/openedx-app-ios/assets/128456094/6d17a0c3-a6fa-4a14-b204-e3994c88bb1b

## Empty State

Screenshot 2024-05-10 at 16 19 55

## Horizontal Position

https://github.com/openedx/openedx-app-ios/assets/128456094/9d6e4376-62a1-453f-abba-2b61f04738c7

## Feature enabled by default.
🚨 You should use key:
```
DASHBOARD:
       TYPE: "list"
```
to enable a list style Dashboard (old style)

## API
Since the new APIs are not available in the master branch, please use the sandbox:
```
API_HOST_URL: 'https://axim-mobile.raccoongang.net'
OAUTH_CLIENT_ID: 'zP3vPz00c8fTRpYjNbVSlA1fxt9LnCxTM4JK1KQ0'
```

### CODE REVIEW

**Comment by IvanStepanok** [comment]
Hi @sdaitzman, and thanks for paying attention to the details! I appreciate it🙌

✅ 1. In light and dark mode, the Resume course quick action doesn’t need a border separating it from the other actions. 
✅ 2. In dark mode, I think the View All Courses card may be a bit too dark/difficult to distinguish from the background.
✅ (2) The account settings gear in the top right should switch to the updated icon

Simulator Screenshot - iPhone 15 - 2024-05-13 at 23 42 59 copy Simulator Screenshot - iPhone 15 - 2024-05-13 at 23 43 04 copy

✅ 3. In the All Courses view in light mode, the filter options currently use a light gray border color, and keep that border color when selected. 
✅ 4. Similarly, in dark mode, course filter options should have no border and should just change their background from Half Dark to Accent when selected.

Simulator Screenshot - iPhone 15 - 2024-05-13 at 23 30 31 copy Simulator Screenshot - iPhone 15 - 2024-05-13 at 23 30 33 copy

✅ 5. When the Programs/Courses dropdown switcher is tapped, the dropdown icon should animate 

https://github.com/openedx/openedx-app-ios/assets/128456094/a09acf1b-5a18-4653-b65d-71d6362839ce

⛔️(3) Selected dashboard-level tabs in dark mode should switch over to using the Dark Theme Accent Text color
Changing the AccentColor can affect other parts of the application. I think it will be better to make these changes in another round.
⛔️ (4) Currently, the deep-link from some quick actions like upcoming assignments/resume course jump to the course home.
In these cases, we need to preload the course structure to move forward. We can't skip this step:(

**Comment by sdaitzman** [comment]
Hey @IvanStepanok, thanks so much for these updates! This looks all good to me now, and we can hold (3) and (4) for future improvements

**Review by volodymyr-chekyrta — APPROVED** [review]

**Comment by moiz994** [comment]
referring to point 5 [here](https://github.com/openedx/openedx-app-ios/pull/434#issuecomment-2108802454).

@sdaitzman shouldn't the Course/Program switcher be animated in the reverse order?
It should be pointing down by default which indicates a dropdown and should be pointing up once opened?

**Comment by sdaitzman** [comment]
@moiz994 great catch, thank you! Yes, the Course/Program switcher dropdown icon should be flipped.

CC @IvanStepanok

**Comment by IvanStepanok** [comment]
@sdaitzman @moiz994 
Done 🙌

https://github.com/openedx/openedx-app-ios/assets/128456094/c670aec8-a6e3-48fb-8ccd-4954fe3555d4

**Comment by rnr** [comment]
@IvanStepanok  Maybe we want to scroll through the menu a bit when opening via the "View Assignments" button to see the currently selected item?
@sdaitzman @moiz994  wdyt?

**Comment by sdaitzman** [comment]
@rnr thanks for checking about these options.

For the first question: yes, I think it makes sense to scroll the course header navigation to the active tab when jumping there.

For the second question, I think it depends on the plan for implementing offline course storage improvements. If those views are accessible offline in the future, the shortcuts to them should be as well.

I would lean slightly towards keeping the quick actions for now and showing a message like the "Slow or no internet connection" message that appears on the bottom of course home in that video until the content is available offline in the future. I'm curious what others think about this question though— can bring it up next mobile design weekly call if that would be helpful.

File: Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift
Comment (rnr) [inline Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift:18]: Maybe we could add config to the PrimaryCourseDashboardViewModel in ScreenAssembly to be consistent with other models/views?

**Review by rnr — CHANGES_REQUESTED** [review]

**Comment by IvanStepanok** [comment]
@rnr Thanks for the review! All the suggested changes have been addressed and corrected.🔥

**Review by rnr — APPROVED** [review]
LGTM

**Comment by rnr** [comment]
Just to report: I've added a new issue with the [double refresh of the Dates tab](https://github.com/openedx/openedx-app-ios/issues/438) (it's not related to this PR).

**Review by saeedbashir — CHANGES_REQUESTED** [review]
* The drop down menu should hide itself on tapping anywhere on screen. It's not hiding even on going to course home.
* The primary course is not updating on dashboard screen when learner goes to a different course and course updated for primary enrollment. 
* I'm not sure how the search is planned on View All Course Screen, but I guess it should only search course from the enrolled courses and even may respect the selected filter.
* The dates page gets loaded twice when we go to dates page directly from dashboard


https://github.com/openedx/openedx-app-ios/assets/5606473/61ff4b70-889e-4262-80f3-ff6d0fe9b6de


https://github.com/openedx/openedx-app-ios/assets/5606473/e02ddbbc-9ebc-45b7-8de1-f21fad58f17e


https://github.com/openedx/openedx-app-ios/assets/5606473/bd67ec42-88b4-42c1-814c-4f5b435e9add

File: Dashboard/Dashboard/Presentation/AllCoursesViewModel.swift
Comment (IvanStepanok) [reply Dashboard/Dashboard/Presentation/AllCoursesViewModel.swift]: Saeed, I've updated the code to move self.nextPage += 1 outside the if-else block. This simplifies the flow and ensures that nextPage is incremented regardless of the condition, avoiding redundant code and ensuring consistency in pagination.

File: Core/Core/Data/Model/Data_PrimaryEnrollment.swift
Comment (IvanStepanok) [reply Core/Core/Data/Model/Data_PrimaryEnrollment.swift]: ✅

File: Core/Core/Data/Model/Data_PrimaryEnrollment.swift
Comment (IvanStepanok) [reply Core/Core/Data/Model/Data_PrimaryEnrollment.swift]: Saeed, while I understand the idea of bundling common properties into a shared object, there is a potential risk that the models might evolve and diverge in the future. For now, let's defer this decision until we have a clearer understanding of the long-term requirements.

File: Core/Core/Data/Model/Data_PrimaryEnrollment.swift
Comment (IvanStepanok) [reply Core/Core/Data/Model/Data_PrimaryEnrollment.swift]: Make sense👍

File: Core/Core/Data/Model/Data_PrimaryEnrollment.swift
Comment (IvanStepanok) [reply Core/Core/Data/Model/Data_PrimaryEnrollment.swift:112]: Saeed, since there are no other Assignment structs in the codebase, there isn't a need to rename it to CourseAssignment. Keeping it as Assignment keeps the code simpler and avoids unnecessary verbosity. If we encounter a need for another Assignment struct in the future, we can revisit this decision and consider renaming at that time.

File: Course/Course/Presentation/Container/CourseContainerViewModel.swift
Comment (IvanStepanok) [reply Course/Course/Presentation/Container/CourseContainerViewModel.swift]: sure👌

**Comment by IvanStepanok** [comment]
@saeedbashir Thanks for review!

#### 1. The drop down menu should hide itself on tapping anywhere on screen. It's not hiding even on going to course home.
I have implemented this change as requested. It was a valid point.

#### 2. The primary course is not updating on the dashboard screen when the learner goes to a different course and the course is updated for primary enrollment.
This issue has been fixed.

#### 3. I'm not sure how the search is planned on the View All Course Screen, but I guess it should only search courses from the enrolled courses and even may respect the selected filter.
Currently, we are using the old search screen. The logic for this feature will be defined in a future scope.

#### 4. The dates page gets loaded twice when we go to the dates page directly from the dashboard.
@rnr has created a new issue for this bug, as mentioned earlier.

**Comment by sdaitzman** [comment]
Hi @IvanStepanok, on the third question, I think it would make sense to hide the search feature entirely on the View All Courses screen over reusing the all-catalog search there. A user initiating a search from that page is very likely intending to search only enrolled courses. Since that's out of scope for current development, I think it makes more sense to hide the feature entirely rather than search the full catalog.

The vast majority of users will be able to find all their courses without using search, especially by using the filters.

File: Dashboard/Dashboard/Data/Network/DashboardEndpoint.swift
Comment (saeedbashir) [reply Dashboard/Dashboard/Data/Network/DashboardEndpoint.swift:22]: ok

**Comment by saeedbashir** [comment]
@IvanStepanok @volodymyr-chekyrta Is it documented that how a course will be moved to the primary course? I'm trying to figure it out but I'm not getting the logic. I tried by going through a course which isn't a primary course and I've completed multiple components of that course but still the course wasn't moved to the primary course.

File: Course/Course/Domain/CourseInteractor.swift
Comment (saeedbashir) [inline Course/Course/Domain/CourseInteractor.swift:67]: Wouldn't this be an overhead by refreshing enrollments on each component completed? 
What do you think about about the approach, mark it that enrollments needs to be refreshed and refresh the enrollments when user move to either course home or dashboard screen?

**Review by saeedbashir — CHANGES_REQUESTED** [review]
* As Sam suggested, please remove search from View All Course page as it's misleading / not completed.
* I'm moving to the course home and facing an issue in API. Then there isn't a going back. We might create an issue if it doesn't fall under the scope of current PR. Please see attached video for reference
* It seems the touch is being ignored when I click on course to go to the course home. Ideally, the touch shouldn't be ignored. The touch will do the action and close the opened menu. And if the touch location is on the empty screen / non actionable then it will close the menu. Please see attached video for reference

https://github.com/openedx/openedx-app-ios/assets/5606473/572c01bd-caf8-4134-9226-e3e181269526

https://github.com/openedx/openedx-app-ios/assets/5606473/7a975336-ebff-4257-bd33-a174db514927

File: Core/Core/Data/Model/Data_PrimaryEnrollment.swift
Comment (saeedbashir) [reply Core/Core/Data/Model/Data_PrimaryEnrollment.swift:112]: For the time being we can go with this approach. The naming seems a bit odd from the other supporting structs names.

**Comment by IvanStepanok** [comment]
@saeedbashir 🙌

As Sam suggested, please remove search from View All Course page as it's misleading / not completed.
✅ I have removed the search bar as requested.

I'm moving to the course home and facing an issue in API. Then there isn't a going back. We might create an issue if it doesn't fall under the scope of current PR. 
✅ Yes, we will need to address this in a separate scope.

It seems the touch is being ignored when I click on course to go to the course home. Ideally, the touch shouldn't be ignored. The touch will do the action and close the opened menu.

⚠️ Unfortunately, I haven't been able to come up with a better solution for this issue. If you have any suggestions, they would be greatly appreciated.

Wouldn't this be an overhead by refreshing enrollments on each component completed?
✅ Thank you for pointing that out. I have adjusted the logic so that enrollments are now refreshed only once.

**Comment by IvanStepanok** [comment]
@saeedbashir 
About this:
It seems the touch is being ignored when I click on course to go to the course home. Ideally, the touch shouldn't be ignored. The touch will do the action and close the opened menu.

I conducted a little investigation, and from what I observed in various software, this behavior is standard. The first click closes the open dropdown menu, and the second click performs the intended action.

**Comment by IvanStepanok** [comment]
@sdaitzman

I investigated the behavior of dropdown menus in various applications. Consistently, the first click outside an open dropdown closes it, and the second click performs the intended action.

I've attached a video showing this behavior in multiple production apps:

https://github.com/openedx/openedx-app-ios/assets/128456094/f6e017a5-b63d-4674-ae45-8cfbc1bd555e


https://github.com/openedx/openedx-app-ios/assets/128456094/cd7f708c-c7d1-4f74-ab49-8c66f1259895

https://github.com/openedx/openedx-app-ios/assets/128456094/a7e3bc05-821d-48a6-9cdb-b3f5e1fe514c


https://github.com/openedx/openedx-app-ios/assets/128456094/b37fd344-3a3e-471d-8b24-7514d706402b



https://github.com/openedx/openedx-app-ios/assets/128456094/aa843dc2-a7f7-45a9-bce8-0cc680ed0391



This behavior is expected for several reasons:

User Expectation: Users expect that clicking outside a dropdown will close it.
Preventing Accidental Actions: This prevents unintended actions from occurring.
Consistency: It aligns with user experience across different applications.
This standard behavior enhances usability by providing a predictable interface. Please review the video and consider these points.

**Comment by sdaitzman** [comment]
Hey @IvanStepanok @saeedbashir— I agree that the expected behavior when tapping outside an open menu is to close the menu, not to perform another action

File: Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift
Comment (saeedbashir) [inline Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift:188]: This isn't working as expected. I guess the right way is to set this variable true on receiving the notification `onblockCompletionRequested` and then call `updateEnrollmentsIfNeeded` when view appears.

File: Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift
Comment (saeedbashir) [inline Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift:283]: I believe this is an accidental comment. PLease revert it.

**Review by saeedbashir — CHANGES_REQUESTED** [review]
The settings button and tab switcher isn't closing the menu.

https://github.com/openedx/openedx-app-ios/assets/5606473/e25e8c9f-c3f5-4f95-bbd9-b3025c404aeb

File: Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift
Comment (IvanStepanok) [reply Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift:283]: Thanks! 🙏

File: Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift
Comment (saeedbashir) [reply Dashboard/Dashboard/Presentation/PrimaryCourseDashboardView.swift:188]: The solution is working perfectly fine, but the issue is that it's a bit confusing. The app is launched and the enrollment API is just hit, and because of this part, the code is saying an update is still needed.

In ideal scenario, the `updateNeeded` only marked true when there is a need. Like if the completion call is called for a course that is not a primary course already.

**Comment by saeedbashir** [comment]
> @IvanStepanok @volodymyr-chekyrta Is it documented that how a course will be moved to the primary course? I'm trying to figure it out but I'm not getting the logic. I tried by going through a course that isn't a primary course, and I've completed multiple components of that course, but still, the course wasn't moved to the primary course.
> 
> Sometimes the same approach works, and sometimes it doesn't.

@sdaitzman is it documented somewhere? It will be helpful for us, especially for QA team, to understand the feature and then write the test cases around it.

**Comment by saeedbashir** [comment]
> @IvanStepanok @volodymyr-chekyrta Is it documented that how a course will be moved to the primary course? I'm trying to figure it out but I'm not getting the logic. I tried by going through a course that isn't a primary course, and I've completed multiple components of that course, but still, the course wasn't moved to the primary course.
> 
> Sometimes the same approach works, and sometimes it doesn't.

@sdaitzman Is it possible for you to share details of the story https://youtrack.raccoongang.com/issue/AXM-506. 
I'm going through the content of a non-primary course and the completion API has been calling for multiple components, but still, I'm getting the old primary course in the enrollments.

**Comment by saeedbashir** [comment]
@IvanStepanok In the list dashboard view, the Programs tab has disappeared. Please make sure it will also be configured the old way (in the bottom tab bar).

image

**Comment by IvanStepanok** [comment]
> @IvanStepanok In the list dashboard view, the Programs tab has disappeared. Please make sure it will also be configured the old way (in the bottom tab bar).
> 

Thank you for bringing this to my attention. I have addressed the issue, and the Programs tab has been restored to the bottom tab bar as before.

**Comment by volodymyr-chekyrta** [comment]
> > @IvanStepanok @volodymyr-chekyrta Is it documented that how a course will be moved to the primary course? I'm trying to figure it out but I'm not getting the logic. I tried by going through a course that isn't a primary course, and I've completed multiple components of that course, but still, the course wasn't moved to the primary course.
> > Sometimes the same approach works, and sometimes it doesn't.
> 
> @sdaitzman Is it possible for you to share details of the story https://youtrack.raccoongang.com/issue/AXM-506. I'm going through the content of a non-primary course and the completion API has been calling for multiple components, but still, I'm getting the old primary course in the enrollments.
> 
> cc @KyryloKireiev

@saeedbashir Thank you for your review!

Let’s move forward with this PR so we can continue developing the related features.

It’s indeed odd that you don’t see a new primary course after xBlock completion; this might be a sandbox issue. However, this is not within the iOS area. All back-end issues will be addressed in the back-end PR, and we will cover any minor issues by the next PRs.

For now PR looks good for merging to me.

@GlugovGrGlib FYI sandbox issue.

cc @marcotuts @e0d @miankhalid

**Comment by saeedbashir** [comment]
> Let’s move forward with this PR so we can continue developing the related features.

@volodymyr-chekyrta The PR isn't block because of this issue/requirements, I just needed documentation of this so that QA can update the test document.

**Comment by volodymyr-chekyrta** [comment]
> > Let’s move forward with this PR so we can continue developing the related features.
> 
> 
> 
> @volodymyr-chekyrta The PR isn't block because of this issue/requirements, I just needed documentation of this so that QA can update the test document. 

I'll provide it asap

**Comment by volodymyr-chekyrta** [comment]
> > Let’s move forward with this PR so we can continue developing the related features.
> 
> @volodymyr-chekyrta The PR isn't block because of this issue/requirements, I just needed documentation of this so that QA can update the test document.

@saeedbashir, please find the feature documentation below

**Dashboard Level Navigation: Primary course**
This feature allows the mobile application to display user course enrollments, prioritizing the most relevant course information. 
The application shows a primary course card, secondary course cards, and provides an option to view all courses. 
This feature ensures users can easily access and manage their current course progress.

**User Story**
As a user of the mobile application,
I want to see my most relevant course enrollment details prioritized,
so that I can quickly access and manage my current course progress efficiently.

**Acceptance Criteria:**
1.	Empty State:
	•	When I have no course enrollments, the screen should show an empty state, indicating that there are no courses to display.
2.	Primary and Secondary Cards:
	•	When I open the Learn screen, I should see a Primary card, a Secondary card (if applicable), and a “View All” button for easy navigation.
3.	Primary Card Information:
	•	The Primary card should display the following information to help me quickly understand my current course status:
	•	Organization name
	•	Course title
	•	Course date
	•	Information about any past due assignments
	•	Details of the next due assignment
4.	Secondary Card Restrictions:
	•	If any course has access issues, I should see a lock icon indicating restricted access.
5.	Navigation to All Courses:
	•	By clicking the “View All” button, I should be redirected to the All Courses screen where I can see a list of all my courses with filters(categories All courses | In progress | Completed | Expired) and a search button.
6.	All Courses Page Details:
	•	On the All Courses page, each course entry should display:
	•	My progress in the course
	•	The course date
	•	The course title
	•	A lock icon if there are any access restrictions
7.	Offline Access:
	•	I should be able to access the screen with the primary course information even when I am offline, ensuring continuity in tracking my course progress.

**Comment by saeedbashir** [comment]
> 7\. Offline Access:
>     •	I should be able to access the screen with the primary course information even when I am offline, ensuring continuity in tracking my course progress.

@volodymyr-chekyrta Is offline support also part of this PR. I'm asking this because it's not working.
image

**Comment by volodymyr-chekyrta** [comment]
> > 7. Offline Access:
> > •	I should be able to access the screen with the primary course information even when I am offline, ensuring continuity in tracking my course progress.
> 
> @volodymyr-chekyrta Is offline support also part of this PR. I'm asking this because it's not working.

@saeedbashir, the documentation may sound a bit confusing, but it's about the "Primary Course" screen, not the "All Courses" screen.

Offline Access:
• I should be able to access the screen with the `primary course` information even when I am offline, ensuring continuity in tracking my course progress.

**Comment by volodymyr-chekyrta** [comment]
> > 7. Offline Access:
> > •	I should be able to access the screen with the primary course information even when I am offline, ensuring continuity in tracking my course progress.
> 
> @volodymyr-chekyrta Is offline support also part of this PR. I'm asking this because it's not working.

@saeedbashir, Yes, according to the documentation, the primary course screen should be cached.

@IvanStepanok could you double-check what happened with caching?

**Comment by IvanStepanok** [comment]
@saeedbashir, @volodymyr-chekyrta,

I've made changes to handle the offline access issue by ensuring that the CoreData model is updated correctly. The main changes include:

Clearing Old Data: Before saving new enrollments, all existing data in the CoreData model is deleted. This approach avoids potential migration issues and ensures data consistency.
Saving New Data: The new data is then saved into CoreData.
These changes should help in caching the primary course screen correctly, allowing access to the course information even when offline.

It's worth considering applying a similar approach to other parts of the project that interact with CoreData to ensure overall data consistency and integrity. We should review the rest of the project to identify and address any similar issues.

**Comment by saeedbashir** [comment]
@IvanStepanok Great! the offline mode is now working. Now the last item remaining in this PR is https://github.com/openedx/openedx-app-ios/pull/434#issuecomment-2130095030 after that we will be good to merge this PR.

**Comment by IvanStepanok** [comment]
> @IvanStepanok I'm not sure why but the program webview is not loading with course dashboard view. I've used the v3 of enrollments so the enrollments screen is not loading (any how this can happen anytime).
> 
> I know programs are not available on your intances, what you can do is you can use any dummy URL for programs, e.g `www.courses.edx.org` and make sure it will display under the programs menu. Even the empty course screen is not hidden when switching to the program view.
> 
> Another issue is, when I use pull to refresh the programs view, it calls the enrollments API instead of reloading the programs webview. The program's webview is working fine in list view.
> 
>  Screen.Recording.2024-05-24.at.10.38.29.PM.mov

As an example, I try to add a base URL for Programs. They run the redirect flow, but if we comment out the block about redirecting, the programs view loads fine. Please check it on your side.🙌
Screenshot 2024-05-29 at 17 47 22

**Comment by saeedbashir** [comment]
@IvanStepanok I tried to capture the behaviour in the following screen recording. Please have a look, and if something isn't clear, we can connect on Slack.

https://github.com/openedx/openedx-app-ios/assets/5606473/b5bf585d-e629-413b-b243-6e597cff61d2

**Review by saeedbashir — DISMISSED** [review]
LGTM 👍

**Review by volodymyr-chekyrta — APPROVED** [review]

**Comment by openedx-webhooks** [comment]
@IvanStepanok 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-app-ios&entry.867055334=https://github.com/openedx/openedx-app-ios/pull/434&entry.1484655318=https://github.com/IvanStepanok&entry.752974735=2024-05-09+12:31&entry.1917517419=2024-06-01+13:37&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**
